### PR TITLE
[R] Basic implementation for R JSON serialization.

### PR DESCRIPTION
**Pull request by trivialfis**
* Change `xgb.save.raw' into full serialization instead of simple model.
* Add `xgb.load.raw' for unserialization.
* Add `xgb.config' getter and setter.

This might be breaking change as previously the `xgb.save.raw` function strips out some parameters due to binary IO.  Now it's a full serialization of all internal states.

### CODE REVIEW

**Comment by hcho3** [comment]
@trivialfis Do you need help with this PR?

**Comment by trivialfis** [comment]
Will get back to this nest week.  Currently traveling.

**Comment by LeZhengThu** [comment]
Hi, any progress on this PR? I learned that the JSON IO is only available in Python, not in R, right?

**Comment by LeZhengThu** [comment]
Hi, any progress on this PR? I learned that the JSON IO is only available in Python, not in R, right?

**Comment by trivialfis** [comment]
It's on R, you can pass a file name with JSON extension to obtain it.  This PR handlea saving the internal configuration, not the model.

File: R-package/tests/testthat/test_callbacks.R [hunk 1]
```diff
@@ -184,15 +184,19 @@ test_that("cb.save.model works as expected", {
   expect_equal(xgb.ntree(b1), 1)
   b2 <- xgb.load('xgboost_02.model')
   expect_equal(xgb.ntree(b2), 2)
+
+  xgb.config(b2) <- xgb.config(bst)
+  expect_equal(xgb.config(bst), xgb.config(b2))
```
Comment (hcho3) [inline R-package/tests/testthat/test_callbacks.R:189]: We should also test `xgb.serialize` and `xgb.unserialize`, and see whether they preserve configuration.

File: R-package/CMakeLists.txt [hunk 2]
```diff
@@ -32,3 +32,7 @@ set_target_properties(
 set(XGBOOST_DEFINITIONS "${XGBOOST_DEFINITIONS};${R_DEFINITIONS}" PARENT_SCOPE)
 set(XGBOOST_OBJ_SOURCES $<TARGET_OBJECTS:xgboost-r> PARENT_SCOPE)
 set(LINKED_LIBRARIES_PRIVATE ${LINKED_LIBRARIES_PRIVATE} ${LIBR_CORE_LIBRARY} PARENT_SCOPE)
+
+if (USE_OPENMP)
+  target_link_libraries(xgboost-r PRIVATE OpenMP::OpenMP_CXX)
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply R-package/CMakeLists.txt:37]: @hcho3 I saw a warning about unknown omp pragma.  `xgboost-r` is a different CMake target here.   Maybe I need to do a bigger refactor that we can gather all source file names into one CMakeLists.

File: R-package/tests/testthat/test_callbacks.R (hunk 1 above)
Comment (trivialfis) [reply R-package/tests/testthat/test_callbacks.R:189]: Added test.

**Comment by trivialfis** [comment]
@hcho3  R has a builtin function called `serialize`, I'm not sure how to integrate with it with class method. (take `predict` as an example).  Also there seems to be a lots of serialization methods in R ...

**Comment by hcho3** [comment]
@trivialfis We are already integrating with `saveRDS()`. This should be good enough.

**Comment by trivialfis** [comment]
@hcho3 I didn't notice any integration with RDS?

File: R-package/tests/testthat/test_basic.R [hunk 3]
```diff
@@ -219,6 +219,21 @@ test_that("training continuation works", {
...
 })
 
+test_that("model serialization works", {
+  out_path <- "model_serialization"
+  dtrain <- xgb.DMatrix(train$data, label = train$label)
+  watchlist = list(train=dtrain)
+  param <- list(objective = "binary:logistic")
+  booster <- xgb.train(param, dtrain, nrounds = 4, watchlist)
+  raw <- xgb.serialize(booster)
+  saveRDS(raw, out_path)
```
Comment (hcho3) [inline R-package/tests/testthat/test_basic.R:229]: @trivialfis This test is using `saveRDS`.

**Comment by trivialfis** [comment]
@hcho3 that's just saving a raw vector returned by xgboost as RDS.  I wouldn't call it integration.  I checked how tensorflow does serialization on R, didn't find any integration either.  So I will merge it looks good to you.

**Comment by trivialfis** [comment]
Phew .. Finally got all the initial JSON code merged.

**Comment by hcho3** [comment]
@trivialfis It's been a long journey for us. I appreciate your hard work.

**Pull request merged**